concourse/atc
# Use Kubernetes secrets as a credential manager

Address PR feedback in #189. 

- Lookup secrets with the same heuristic used in Vault.
- Use a prefix for namespaces to avoid collisions. (default: concourse-)

Testing locally:
1. Get `kubectl` to target a cluster.
1. Edit concourse/dev/atc (or concourse/dev/atc-dockerdb) and add `--kubernetes-config-path=[userhome]/.kube/config` to the command line flags.
1. `./dev/start`
1. `fly -t lite sp -p hello-world -c hello-world.yml` with the following pipeline:

    ```
    jobs:
    - name: hello-world
      plan:
      - task: say-hello
        config:
          platform: linux
          image_resource:
            type: docker-image
            source: {repository: alpine}
          params:
            SECRET: ((my-secrets.secret1))
          run:
            path: /bin/sh
            args: ["-c", "echo $SECRET"]

1. Run the pipeline and see an error in the UI `Expected to find variables: my-secrets`. Also see a log message `k8s-secret-not-found` with data about the namespace and secrets that were searched for. 
1. `kubectl create namespace concourse-main`
1. `kubectl config set-context $(kubectl config current-context) --namespace=concourse-main`
1. `kubectl create secret generic hello-world.my-secrets --from-literal "secret1=secretvalue1"`
1. Run the pipeline and see `secretvalue1` printed.
1. `kubectl delete secret  hello-world.my-secrets`
1. `kubectl create secret generic my-secrets --from-literal "secret1=secretvalue2"`
1. Run the pipeline and see `secretvalue2` printed.
1. `kubectl create secret generic my-secret-value --from-literal "value=secretvalue3"`
1. Change `((my-secrets.secret1))` to `((my-secret-value))` in the pipeline.
1. Run the pipeline and see `secretvalue3` printed.
